benchmarks/src/bench/json_multi_export.rs

```diff
@@ -4,106 +4,117 @@ use std::{
     path::PathBuf,
 };
 
-use mongodb::{Client, Collection, Database};
+use anyhow::Result;
+use futures::stream::{FuturesUnordered, StreamExt, TryStreamExt};
+use mongodb::{bson::doc, Client, Collection, Database};
 
-use crate::{
-    bench::{parse_json_file_to_documents, Benchmark, COLL_NAME, DATABASE_NAME},
-    error::Result,
-};
+use crate::bench::{parse_json_file_to_documents, Benchmark, COLL_NAME, DATABASE_NAME};
 
 const TOTAL_FILES: usize = 100;
 
 pub struct JsonMultiExportBenchmark {
     db: Database,
     coll: Collection,
-    num_threads: usize,
 }
 
 // Specifies the options to a `JsonMultiExportBenchmark::setup` operation.
 pub struct Options {
-    pub num_threads: usize,
     pub path: PathBuf,
     pub uri: String,
 }
 
+#[async_trait::async_trait]
 impl Benchmark for JsonMultiExportBenchmark {
     type Options = Options;
 
-    fn setup(options: Self::Options) -> Result<Self> {
-        let client = Client::with_uri_str(&options.uri)?;
+    async fn setup(options: Self::Options) -> Result<Self> {
+        let client = Client::with_uri_str(&options.uri).await?;
         let db = client.database(&DATABASE_NAME);
-        db.drop()?;
+        db.drop(None).await?;
 
         let coll = db.collection(&COLL_NAME);
 
+        let mut tasks = FuturesUnordered::new();
+
         for i in 0..TOTAL_FILES {
-            let json_file_name = options.path.join(format!("ldjson{:03}.txt", i));
-            let file = File::open(&json_file_name)?;
+            let path = options.path.clone();
+            let coll = coll.clone();
+
+            tasks.push(async move {
+                let json_file_name = path.join(format!("ldjson{:03}.txt", i));
+                let file = spawn_blocking_and_await!(File::open(&json_file_name))?;
+
+                let docs = spawn_blocking_and_await!(parse_json_file_to_documents(file))?;
 
-            let docs = parse_json_file_to_documents(file)?;
+                for mut doc in docs {
+                    doc.insert("file", i as i32);
+                    coll.insert_one(doc, None).await?;
+                }
+
+                let ok: anyhow::Result<()> = Ok(());
+                ok
+            });
+        }
 
-            for mut doc in docs {
-                doc.insert("file", i as i32);
-                coll.insert_one(doc, None)?;
-            }
+        while let Some(result) = tasks.next().await {
+            println!("done!");
+            result?;
         }
 
-        Ok(JsonMultiExportBenchmark {
-            db,
-            coll,
-            num_threads: options.num_threads,
-        })
+        Ok(JsonMultiExportBenchmark { db, coll })
     }
 
-    fn do_task(&self) -> Result<()> {
-        let mut num_each = TOTAL_FILES / self.num_threads;
-        let remainder = TOTAL_FILES % self.num_threads;
-        if remainder != 0 {
-            num_each += 1;
-        }
+    async fn do_task(&self) -> Result<()> {
+        let mut tasks = FuturesUnordered::new();
```

Comment:
Previously, this benchmark spawned a number of threads manually and divided up the work among them equally. Now that the driver is async, it's more performant to just spawn a bunch of tasks and then let the runtime do the work of dividing them in a performant manner. Running this one manually a spawnhost showed promising performance gains from this switch, although I'll wait until we can actually verify the results in our automatic runs before doing any victory laps.